src/backend/lib/utils.js

```diff
@@ -0,0 +1,299 @@
+// @ts-check
+const semver = require("semver");
+const hasha = require("hasha");
+const path = require("path");
+const fs = require("fs");
+const stream = require("stream");
+const { promisify } = require("util");
+const { once } = require("events");
+const pDefer = require("p-defer");
+const copyError = require("utils-copy-error");
+const ApkParser = require("app-info-parser/src/apk");
+const ZipFs = require("@gmaclennan/zip-fs");
+
+/** @typedef {import('./types').InstallerInt} InstallerInt */
+/** @typedef {import('./types').InstallerExt} InstallerExt */
+
+module.exports = {
+  installerCompare,
+  isUpgradeCandidate,
+  getBestUpgradeCandidate,
+  getInstallerInfo,
+  beforeAfterStream,
+  stringifyInstaller,
+};
+
+/**
+ * Compare two installer options, returns -1 if a is less than b, 0 if same, 1
+ * if a is greater than b
+ *
+ * @param {InstallerInt | InstallerExt} a
+ * @param {InstallerInt | InstallerExt} b
+ * @returns {-1 | 0 | 1}
+ */
+function installerCompare(a, b) {
+  const aIsValidSemver = semver.valid(a.versionName);
+  const bIsValidSemver = semver.valid(b.versionName);
+  if (!aIsValidSemver && !bIsValidSemver) return 0;
+  if (!aIsValidSemver) return -1;
+  if (!bIsValidSemver) return 1;
+  // Important to validate semver before here, otherwise this will throw
+  const semverCompare = semver.compare(a.versionName, b.versionName);
+  if (semverCompare !== 0) return semverCompare;
+  if (a.versionCode < b.versionCode) return -1;
+  if (a.versionCode > b.versionCode) return 1;
+  return 0;
+}
+
+/**
+ * Check if an installer is compatible with the given device. Requirements:
+ *
+ * 1. Must support arch of device
+ * 2. Must support SDK of device
+ * 3. Version code must be greater or equal to currently installed app
+ * 4. Installer must be for platform 'android'
+ * 5. applicationId must match applicationId of currently installed app
+ * 6. versionName must be valid semver
+ *
+ * @param {object} options
+ * @param {import('./types').DeviceInfo} options.deviceInfo
+ * @param {InstallerInt | InstallerExt} options.installer
+ * @param {InstallerInt | InstallerExt} options.currentApkInfo
+ * @returns {boolean}
+ */
+function isUpgradeCandidate({ deviceInfo, installer, currentApkInfo }) {
+  // Filter out the current APK
+  if (installer.hash === currentApkInfo.hash) return false;
+  // TODO: 64 bit devices can run 32 bit APKs
+  const isSupportedAbi = !!installer.arch.find(arch =>
```

Comment:
Just a little simpler.
```suggestion
  const isSupportedAbi = installer.arch.some(arch =>
```